content/blog/2019-09-25-bril-type-inference.md

```diff
@@ -0,0 +1,204 @@
+type `int`; otherwise we'll throw a type unification error. Finally, we do
+the same for the destination, making sure that it is typed as `bool`.
+
+For our implementation, we simply iterate through each instruction and determine
+the types of our variables. The one case where this presents issues is with
+`id`. In general, `id` sets the type of the destination to the type of the
+variable on the right hand side. However, what happens if we don't know the type
+of the variable being copied? For example, consider this Bril program:
+```
+main {
+  jmp later;
+earlier:
+  x = id y;
+  ret;
+later:
+  y = const 5;
+  jmp earlier;
+}
+```
+This is a valid program that should typecheck. Specifically, `x` and `y` are
+both ints. However, if we naively go sequentially through each instruction, we
+don't know the type of `y` until we have `y = const 5`, at which point it is too
+late to type `x`. To resolve this, we have two options:
+1. Rerun the type inference algorithm, stopping when no more variables have been
+inferred
+2. Keep track of which variables must have the same type, and after 1 pass,
+setting their types to be the same.
+
+For simplicity sake, we choose (1). This means in the worst case, type inference
+will take `O(n^2)` time, where `n` is the number of instructions. This happens
+when there are multiple such `id` assignments as in the example above.
+
+### Type Checking
+With type inference, type checking is relatively simple. After type inference,
+we have the original Bril program and the fully typed Bril program. We then go
+through the original Bril program and make sure that for any variable that has a
+type annotation, the type matches the inferred type. For completeness, I also
+check to make sure that variables aren't being used as labels, and vice versa.
+
+## Evaluation
+To properly evaluate that everything presented here is correct, we have to test
+the parser and the type inferrer. To do these simultaneously, we can run tests
+as follows:
+```
+command = "cat {filename} | bril2json | python ../../infer.py | bril2txt"
+```
+We are taking Bril text programs, turning them into JSON, generating a new
+equivalent typed program, and turning it back to text. First I started with
+simple Bril programs that already existed to build some confidence. Then to gain
+full confidence, I wrote tests that use every kind of operations, e.g.
+arithmetic ops, comparison ops, logical ops, effect ops, and misc. ops.
+
+I tested for both positive and negative results. In other words, I ensured
+Bril programs that *should* typecheck were correctly type-inferred, and programs
+that *shouldn't* typecheck in fact could not be type-inferred. For example:
+```
+main {
+  F1 = const false;
+  T1 = const 1;
+  b1 = and F1 T1;
+}
+```
+Unfortunately I couldn't get Turnt to validate the error message that was
+output, so instead I manually checked to see that the error was what I expected.
```

Comment:
FWIW, this capability was recently added: https://github.com/cucapra/turnt/pull/9

Reply:
Awesome! This is a pretty useful feature.